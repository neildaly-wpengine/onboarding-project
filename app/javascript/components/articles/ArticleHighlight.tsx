import {
  Avatar,
  Button,
  Card,
  CardActionArea,
  CardActions,
  CardContent,
  CardHeader,
  CardMedia,
  IconButton,
  Menu,
  MenuItem,
  Typography,
} from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import FavoriteIcon from "@material-ui/icons/Favorite";
import MoreVertIcon from "@material-ui/icons/MoreVert";
import React, { useState } from "react";
import { Link } from "react-router-dom";
import { createUserInitials } from "../../common/common";
import { Article, AuthenticatedProps } from "../../common/types";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      borderRadius: 10,
      maxWidth: 345,
      margin: 20,
      // display card actions on complete bottom
      display: "flex",
      flexDirection: "column",
      justifyContent: "space-between",
    },
    cardActions: {
      // display card actions on complete bottom
      marginTop: "auto",
    },
    avatar: {
      backgroundColor: theme.palette.primary.main,
    },
    link: {
      textDecoration: "none",
      color: "#000",
    },
  })
);

const ArticleHighlight: React.FC<Article & AuthenticatedProps> = ({
  title,
  content,
  user,
  createdAt,
  id,
  stockImage,
  authenticated,
  currentUserID,
}) => {
  const classes = useStyles();
  const userInitials: string = createUserInitials(user);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const articleOwner: boolean = currentUserID.toString() === user.id;

  const toggleCardHeaderMenu = (e: React.BaseSyntheticEvent): void => {
    setAnchorEl(e.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const menu: JSX.Element =
    authenticated && articleOwner ? (
      <Menu
        id="menu-appbar"
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: "top",
          horizontal: "right",
        }}
        transformOrigin={{
          vertical: "top",
          horizontal: "right",
        }}
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        <MenuItem>
          <Link to={`/articles/edit/${id}`} className={classes.link}>
            Edit
          </Link>
        </MenuItem>
      </Menu>
    ) : (
      <React.Fragment />
    );

  return (
    <Card className={classes.root}>
      <CardHeader
        avatar={
          <Avatar aria-label="user" className={classes.avatar}>
            {userInitials}
          </Avatar>
        }
        action={
          authenticated &&
          articleOwner && (
            <IconButton aria-label="settings" onClick={toggleCardHeaderMenu}>
              <MoreVertIcon />
            </IconButton>
          )
        }
        title={`${user.firstName} ${user.lastName}`}
        subheader={createdAt}
        data-testid="article-card-header"
      />
      {menu}
      <CardActionArea>
        <CardMedia
          component="img"
          alt="blog-image"
          height="200"
          image={stockImage}
          title={title}
        />
        <CardContent>
          <Typography
            gutterBottom
            variant="h5"
            component="h2"
            data-testid="article-card-title"
          >
            {title}
          </Typography>
          <Typography
            variant="body2"
            color="textSecondary"
            component="p"
            data-testid="article-card-content"
          >
            {content.substring(0, 200)}...
          </Typography>
        </CardContent>
      </CardActionArea>
      <CardActions className={classes.cardActions}>
        <IconButton aria-label="add to favorites">
          <FavoriteIcon />
        </IconButton>
        <Button
          size="small"
          color="primary"
          component={Link}
          to={`/articles/${id}`}
          data-testid="article-card-button"
        >
          View More
        </Button>
      </CardActions>
    </Card>
  );
};

export default ArticleHighlight;
