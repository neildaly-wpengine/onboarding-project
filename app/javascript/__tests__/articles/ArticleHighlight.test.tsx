import "@testing-library/jest-dom/extend-expect";
import { cleanup, render, RenderResult } from "@testing-library/react";
import * as React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import ArticleHighlight from "../../components/articles/ArticleHighlight";
import { mockArticle } from "./__helpers__/test-data";

const renderHelper = (): RenderResult => {
  return render(
    <Router>
      <ArticleHighlight
        article={mockArticle}
        auth={{
          authenticated: false,
          currentUserID: "",
        }}
        notifyDelete={jest.fn}
      />
    </Router>
  );
};

afterEach(cleanup);

describe("<ArticleHighlight />", () => {
  describe("Snapshots", () => {
    test("should render an article highlight without crashing", () => {
      const { container } = renderHelper();

      expect(container).toMatchSnapshot();
    });
  });

  describe("Content", () => {
    test("should render the correct title", () => {
      const { getByTestId } = renderHelper();

      expect(getByTestId("article-card-title")).toHaveTextContent(
        mockArticle.title
      );
    });

    test("should render the correct content", () => {
      const { getByTestId } = renderHelper();

      expect(getByTestId("article-card-content")).toHaveTextContent(
        `${mockArticle.content}...`
      );
    });

    test("should render the correct header details", () => {
      const { getByTestId } = renderHelper();

      expect(getByTestId("article-card-header")).toHaveTextContent(
        mockArticle.createdAt
      );
      expect(getByTestId("article-card-header")).toHaveTextContent(
        `${mockArticle.user.firstName} ${mockArticle.user.lastName}`
      );
    });

    test("should render the correct button content", () => {
      const { getByTestId } = renderHelper();

      expect(getByTestId("article-card-button")).toHaveTextContent("View More");
      expect(getByTestId("article-card-button").closest("a")).toHaveAttribute(
        "href",
        `/articles/${mockArticle.id}`
      );
    });
  });
});
