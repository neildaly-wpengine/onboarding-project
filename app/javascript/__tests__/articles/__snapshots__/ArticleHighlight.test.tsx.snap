// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleHighlight /> Snapshots should render an article highlight without crashing 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
      data-testid="article-card-header"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          aria-label="user"
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-3 MuiAvatar-colorDefault"
        >
          TU
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
        >
          Test User
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          2020-06-24T16:10:53.448Z
        </span>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiCardActionArea-root"
      tabindex="0"
      type="button"
    >
      <img
        alt="blog-image"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
        height="200"
        title="Some Fake Title"
      />
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          data-testid="article-card-title"
        >
          Some Fake Title
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          data-testid="article-card-content"
        >
          Some dummy content
          ...
        </p>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiCardActions-root makeStyles-cardActions-2 MuiCardActions-spacing"
    >
      <button
        aria-label="add to favorites"
        class="MuiButtonBase-root MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
        data-testid="article-card-button"
        href="/articles/1"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          View More
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</div>
`;
